### PR TITLE
pin: implement pin/ls with only CoreApi

### PR DESCRIPTION
**Pull request by MichaelMure**
This PR use improvements from https://github.com/ipfs/interface-go-ipfs-core/pull/49 and https://github.com/ipfs/interface-go-ipfs-core/pull/50 to implement `pin ls` only using the CoreApi.

During this refactor I discovered a bug [I introduced in https://github.com/ipfs/go-ipfs/pull/6493, sorry] where the priority between pin type was not respected: when doing a `ipfs pin ls --type=indirect`, direct and recursive pins were not visited first so one of those could have been listed as indirect. This PR also fix this.

### CODE REVIEW

**Comment by MichaelMure** [comment]
CC @Stebalien @magik6k

**Comment by Stebalien** [comment]
#6705 fixes the issue you found.

**Comment by MichaelMure** [comment]
@Stebalien as I understand, there is two solutions here:
1. Change `PinAPI.Ls` to return `(chan coreiface.Pin, chan error)` to match what #6705 did 
2. Keep the interface defined in https://github.com/ipfs/interface-go-ipfs-core/pull/49 and https://github.com/ipfs/interface-go-ipfs-core/pull/50 and return `(<-chan Pin, error)`. In that case, rebasing this PR would essentially be copying my implementation over the one currently in master.

Can you confirm which one you'd like ?

**Comment by Stebalien** [comment]
Let's go with the second. I think we need to evaluate using multiple channels for errors everywhere in the future but that's a larger change.

**Comment by MichaelMure** [comment]
Should be good to go now

**Comment by MichaelMure** [comment]
Someone to merge this one?

**Comment by Stebalien** [comment]
I will look at this when I get a chance (probably not till after the holidays) but I'm going to need to think about the interface changes.

**Comment by MichaelMure** [comment]
Rebased on master. It would be nice to merge that alongside https://github.com/ipfs/interface-go-ipfs-core/pull/50.

**Comment by hsanjuan** [comment]
I'll get back to review this after https://github.com/ipfs/interface-go-ipfs-core/pull/50 is through.

**Comment by MichaelMure** [comment]
@Stebalien do you think this could be included in the 0.5.0 ? That would be neat but it's fine if not.

**Comment by hsanjuan** [comment]
hey @MichaelMure , this is not making it to 0.5.0 (currently very much in feature-freeze). That said these are on our radar for merging afterwards.

**Comment by Stebalien** [comment]
Rebased on master.

File: core/coreapi/pin.go [hunk 1]
```diff
@@ -194,123 +214,106 @@ func (p *pinInfo) Type() string {
...
 
 	keys := cid.NewSet()
-	AddToResultKeys := func(keyList []cid.Cid, typeStr string) error {
+
+	AddToResultKeys := func(keyList []cid.Cid, typeStr string) {
 		for _, c := range keyList {
 			if keys.Visit(c) {
-				select {
-				case ch <- &pinInfo{
+				out <- &pinInfo{
```
Comment (Stebalien) [inline core/coreapi/pin.go:229]: Unless I'm missing something, we still need to select on the context, right?

Comment (MichaelMure) [reply core/coreapi/pin.go:229]: Looks like it. Also, `ctx` should be the first parameter of `pinLsAll`.

Comment (MichaelMure) [reply core/coreapi/pin.go]: Should be good now.

**Pull request merged**